port/cpl_aws.cpp

```diff
@@ -939,9 +939,46 @@ bool VSIS3HandleHelper::GetConfigurationFromEC2(bool bForceRefresh,
                     // Failure: either we are not running on EC2 (or something emulating it)
                     // or this doesn't implement yet IDMSv2
                     // Go on trying IDMSv1
+
+                    // /latest/api/token doesn't work inside a Docker container that
+                    // has no host networking.
+                    // Cf https://community.grafana.com/t/imdsv2-is-not-working-from-docker/65944
+                    if( psResult->pszErrBuf != nullptr &&
+                        strstr(psResult->pszErrBuf, "Operation timed out after") != nullptr )
+                    {
+                        aosOptions.Clear();
+                        aosOptions.SetNameValue("TIMEOUT", "1");
+                        CPLPushErrorHandler(CPLQuietErrorHandler);
+                        CPLHTTPResult* psResult2 =
+                            CPLHTTPFetch( (osEC2RootURL + "/latest/meta-data").c_str(), aosOptions.List() );
+                        CPLPopErrorHandler();
+                        if( psResult2 )
+                        {
+                            if( psResult2->nStatus == 0 && psResult2->pabyData != nullptr )
+                            {
+                                VSIStatBufL sStat;
+                                if( VSIStatL("/.dockerenv", &sStat) == 0 )
+                                {
+                                    CPLDebug("AWS",
+                                             "/latest/api/token EC2 IDMSv2 request timed out, but /latest/metadata succeeded. "
+                                             "Trying with IDMSv1. "
```

Comment:
[Instance Metadata Service (IMDS) v1 is disabled on Amazon Linux 2023](https://docs.aws.amazon.com/linux/al2023/ug/compare-with-al2.html#imdsv2).  When running in a Docker container on an Amazon instance you'll need to update metadata allowing 2 HTTP Hops.  Here's one adhoc way to do that:

```
aws ec2 modify-instance-metadata-options \
    --instance-id <instance_id> \
    --http-put-response-hop-limit 2 \
    --http-endpoint enabled
```